witness/golang/cmd/witness/internal/http/server_test.go

```diff
@@ -0,0 +1,380 @@
+			logIDs:     []string{"bananas", "monkeys"},
+			logPKs:     []string{bPK, mPK},
+			chkpts:     [][]byte{bInit, mInit},
+			wantStatus: http.StatusOK,
+			wantBody:   `["bananas","monkeys"]`,
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			d, closeFn := mustCreateDB(t)
+			defer closeFn()
+			ctx := context.Background()
+			// Set up witness and give it some checkpoints.
+			logs := make([]LogOpts, len(test.logIDs))
+			for i, logID := range test.logIDs {
+				logs[i] = LogOpts{ID: logID,
+					PK:         test.logPKs[i],
+					useCompact: false,
+				}
+			}
+			w := newWitness(t, d, logs)
+			for i, logID := range test.logIDs {
+				if _, err := w.Update(ctx, logID, test.chkpts[i], nil); err != nil {
+					t.Errorf("failed to set checkpoint: %v", err)
+				}
+			}
+			// Now set up the http server.
+			r := mux.NewRouter()
+			server := NewServer(w)
+			server.RegisterHandlers(r)
+			ts := httptest.NewServer(r)
+			defer ts.Close()
+
+			client := ts.Client()
+			url := fmt.Sprintf("%s%s", ts.URL, api.HTTPGetLogs)
+			resp, err := client.Get(url)
+			if err != nil {
+				t.Errorf("error response: %v", err)
+			}
+			if got, want := resp.StatusCode, test.wantStatus; got != want {
+				t.Errorf("status code got != want (%d, %d)", got, want)
+			}
+			if len(test.wantBody) > 0 {
+				body, err := ioutil.ReadAll(resp.Body)
+				if err != nil {
+					t.Errorf("failed to read body: %v", err)
+				}
+				if got, want := string(body), test.wantBody; got != test.wantBody {
+					t.Errorf("got '%s' want '%s'", got, want)
+				}
+			}
+		})
+	}
+}
+
+func TestGetChkpt(t *testing.T) {
+	for _, test := range []struct {
+		desc       string
+		setID      string
+		setPK      string
+		queryID    string
+		queryPK    string
+		c          []byte
+		wantStatus int
+	}{
+		{
+			desc:       "happy path",
+			setID:      "monkeys",
+			setPK:      mPK,
+			queryID:    "monkeys",
+			queryPK:    mPK,
+			c:          mInit,
+			wantStatus: http.StatusOK,
+		}, {
+			desc:       "other log",
+			setID:      "monkeys",
+			setPK:      mPK,
+			queryID:    "bananas",
+			c:          mInit,
+			wantStatus: http.StatusNotFound,
+		}, {
+			desc:       "nothing there",
+			setID:      "monkeys",
+			setPK:      mPK,
+			queryID:    "monkeys",
+			c:          nil,
+			wantStatus: http.StatusNotFound,
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			d, closeFn := mustCreateDB(t)
+			defer closeFn()
+			ctx := context.Background()
+			// Set up witness.
+			w := newWitness(t, d, []LogOpts{{ID: test.setID,
+				PK:         test.setPK,
+				useCompact: false}})
+			// Set a checkpoint for the log if we want to for this test.
+			if test.c != nil {
+				if _, err := w.Update(ctx, test.setID, test.c, nil); err != nil {
+					t.Errorf("failed to set checkpoint: %v", err)
+				}
+			}
+			// Now set up the http server.
+			r := mux.NewRouter()
+			server := NewServer(w)
+			server.RegisterHandlers(r)
+			ts := httptest.NewServer(r)
+			defer ts.Close()
```

Comment:
Is this identical to the setup in the previous test? is it worth extracting to a `createTestEnv(...) (*httptest.Server, func())` method or similar?

Reply:
Okie, done.